File: ArrayList/java/src/ArrayList.java
public class ArrayList<E> {
    private static final int DEFAULT_CAPACITY = 10;
    private static final int DEFAULT_MULTIPLIER = 2;
    private static final double DEFAULT_LOAD_FACTOR = 0.85;
    private static final double DEFAULT_SHRINK_FACTOR = 0.25;

    private E[] array;
    private int size;
    private int capacity;
    private final int multiplier;
    private final double load_factor;
    private final double shrink_factor;

    @SuppressWarnings("unchecked")
    public ArrayList(int capacity, int multiplier, double load_factor, double shrink_factor) {
        if (load_factor > 1 || shrink_factor > 1)
            throw new IllegalArgumentException("Load factor or shrink factor should be less than or equal to 1");
        this.capacity = capacity;
        this.multiplier = multiplier;
        this.load_factor = load_factor;
        this.shrink_factor = shrink_factor;
        this.size = 0;
        this.array = (E[]) new Object[this.capacity];
    }

    public ArrayList() {
        this(DEFAULT_CAPACITY, DEFAULT_MULTIPLIER, DEFAULT_LOAD_FACTOR, DEFAULT_SHRINK_FACTOR);
    }

    private static void checkSize(int index, int size) throws ArrayIndexOutOfBoundsException {
        if (index >= size)
            throw new ArrayIndexOutOfBoundsException();
    }

    public int size() {
        return this.size;
    }

    public boolean isEmpty() {
        return size <= 0;
    }

    private void resize(int capacity) {
        @SuppressWarnings("unchecked") E[] temp = (E[]) new Object[capacity];
        for (int i = 0; i < size; i++) {
            temp[i] = array[i];
        }
        this.array = temp;
    }

    private void grow() {
        if (((double) size / array.length) >= load_factor) {
            this.capacity *= multiplier;
            resize(capacity);
        }
    }

    private void shrink() {
        int n = (int) (shrink_factor * array.length);
        if (size < n) {
            resize(n);
            this.capacity = n;
        }
    }

    private void insert(int index, E element) {
        grow();
        if (index == size) {
            array[size] = element;
        } else {
            for (int i = size; i > index; i--) {
                array[i] = array[i - 1];
            }
            array[index] = element;
        }
        size++;
    }

    public E get(int index) {
        checkSize(index, size - 1);
        return array[index];
    }

    public E set(int index, E element) {
        checkSize(index, size - 1);
        E currentElm = array[index];
        array[index] = element;
        return currentElm;
    }

    public void add(int index, E element) {
        checkSize(index, size + 1);
        insert(index, element);
    }

    /**
     * @param element adds element to the end of the list
     */
    public void add(E element) {
        add(size, element);
    }

    public E remove(int index) {
        checkSize(index, size);
        E element = array[index];
        for (int i = index; i < size; i++) {
            if ((i + 1) == size) {
                array[i] = null;
            } else {
                array[i] = array[i + 1];
            }
        }
        size--;
        shrink();
        return element;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("{");
        for (int i = 0; i < size; i++) {
            builder.append(array[i]);
            if ((i + 1) < size)
                builder.append(", ");
        }
        builder.append("}");
        return builder.toString();
    }

    public static void main(String[] args) {
        ArrayList<Integer> array = new ArrayList<>();
        array.add(1);
        array.add(0, 2);
        System.out.println(array.get(0));
        System.out.println(array);
        for (int i = 0; i < 50; i++) {
            array.add(i + 3);
        }
        for (int i = 0; i < 50; i++) {
            array.remove(0);
        }
        System.out.println(array);
    }
}

File: ArrayList/dart/bin/array_list.dart
void main() {
  ArrayList<int> array = ArrayList();
  array.add(1);
  array.add(2, 0);
  print(array.get(0));
  print(array);
  for (int i = 0; i < 50; i++) {
    array.add(i + 3);
  }
  print(array);
  for (int i = 0; i < 50; i++) {
    array.remove(0);
  }
  print(array);
  print(array.size);
}

class ArrayList<E> {
  static const int DEFAULT_CAPACITY = 10;
  static const int DEFAULT_MULTIPLIER = 2;
  static const double DEFAULT_LOAD_FACTOR = 0.85;
  static const double DEFAULT_SHRINK_FACTOR = 0.25;

  int _size;
  int _capacity;
  List<E?> _array;
  final int _multiplier;
  final double _load_factor;
  final double _shrink_factor;

  ArrayList(
      {int capacity = ArrayList.DEFAULT_CAPACITY,
      int multiplier = ArrayList.DEFAULT_MULTIPLIER,
      double load_factor = ArrayList.DEFAULT_LOAD_FACTOR,
      double shrink_factor = ArrayList.DEFAULT_SHRINK_FACTOR})
      : _capacity = capacity,
        _multiplier = multiplier,
        _load_factor = load_factor,
        _shrink_factor = shrink_factor,
        _size = 0,
        _array = _create_array(capacity) {
    if (load_factor > 1 || shrink_factor > 1) {
      throw Exception(
          "Load factor or shrink factor should be less than or equal to 1");
    }
  }

  static void _checkSize(int index, int size) {
    if (index >= size) {
      throw IndexError.withLength(index, size);
    }
  }

  int get size => _size;

  bool get isEmpty => _size <= 0;

  static List<T?> _create_array<T>(int capacity) {
    return [for (int i = 0; i < capacity; i++) null];
  }

  void _resize(int capacity) {
    List<E?> temp = _create_array(capacity);
    for (int i = 0; i < _size; i++) {
      temp[i] = _array[i];
    }
    this._array = temp;
  }

  void _grow() {
    if ((_size / _array.length) >= _load_factor) {
      this._capacity *= _multiplier;
      _resize(_capacity);
    }
  }

  void _shrink() {
    int n = (_shrink_factor * _array.length).toInt();
    if (_size < n) {
      this._capacity = n;
      _resize(_capacity);
    }
  }

  void _insert(int index, E element) {
    _grow();
    if (index == size) {
      _array[size] = element;
    } else {
      for (int i = _size; i > index; i--) {
        _array[i] = _array[i - 1];
      }
      _array[index] = element;
    }
    _size++;
  }

  E get(int index) {
    _checkSize(index, size - 1);
    return _array[index]!;
  }

  E set(int index, E element) {
    _checkSize(index, size - 1);
    E currentElm = _array[index]!;
    _array[index] = element;
    return currentElm;
  }

  void add(E element, [int? index]) {
    index = index ?? _size;
    _checkSize(index, size + 1);
    _insert(index, element);
  }

  E remove(int index) {
    _checkSize(index, size);
    E element = _array[index]!;
    for (int i = index; i < _size; i++) {
      if ((i + 1) == _size) {
        _array[i] = null;
      } else {
        _array[i] = _array[i + 1];
      }
    }
    _size--;
    _shrink();
    return element;
  }

  @override
  String toString() {
    return [for (E? elm in _array) if (elm != null) elm].toString();
  }
}
